superduperdb/ext/llm/model.py

```diff
@@ -0,0 +1,398 @@
+        self,
+        X: t.Any,
+        y: t.Optional[t.Any] = None,
+        configuration: t.Optional[_TrainingConfiguration] = None,
+        data_prefetch: bool = False,
+        db: t.Optional["Datalayer"] = None,
+        metrics: t.Optional[t.Sequence["Metric"]] = None,
+        select: t.Optional["Select"] = None,
+        validation_sets: t.Optional[t.Sequence[t.Union[str, "Dataset"]]] = None,
+        **kwargs,
+    ):
+        assert configuration is not None, "configuration must be provided"
+
+        training_args = LLMTrainingArguments(**configuration.kwargs)  # type: ignore
+
+        # get device map
+        device_map: t.Union[None, str, t.Dict[str, int]] = None
+        if os.environ.get("LOCAL_RANK") is not None:
+            device_map = {"": int(os.environ.get("LOCAL_RANK", "0"))}
+        elif torch.backends.mps.is_available():
+            device_map = "mps"
+
+        quantization_config = self._create_quantization_config(training_args)
+
+        self.model_kwargs["quantization_config"] = quantization_config
+        self.model_kwargs["device_map"] = device_map
+        self.model, self.tokenizer = self.init_model_and_tokenizer()
+
+        self.tokenizer.model_max_length = (
+            training_args.max_length or self.tokenizer.model_max_length
+        )
+        self._prepare_lora_training(training_args)
+
+        train_dataset, eval_dataset = self.get_datasets(
+            X,
+            y,
+            db,
+            select,
+            data_prefetch=data_prefetch,
+            eval=True,
+            prefetch_size=kwargs.pop("prefetch_size", 10000),
+        )
+
+        # TODO: Defind callbacks about superduperdb side
+        trainer = self.create_trainer(
+            train_dataset,
+            eval_dataset,
+            compute_metrics=metrics,
+            training_args=training_args,
+            **kwargs,
+        )
+        trainer.model.config.use_cache = False
+        trainer.train()
+        trainer.save_state()
+
+    @ensure_initialized
+    def to_call(self, X: t.Any, **kwargs):
+        """
+        Overwrite `Model.to_call` method to support self.object=None.
+        """
+        return self._generate(X, **kwargs)
+
+    def _generate(self, X: t.Any, adapter_name=None, **kwargs):
+        """
+        Private method for `Model.to_call` method.
+        Support inference by multi-lora adapters.
+        """
+        if adapter_name is not None:
+            try:
+                self.model.set_adapter(adapter_name)
+            except Exception as e:
+                raise ValueError(
+                    f"Adapter {adapter_name} is not found in the model, "
+                    "please use add_adapter to add it."
+                ) from e
+
+        elif hasattr(self.model, "disable_adapter"):
+            with self.model.disable_adapter():
+                return self.generate(X, **kwargs)
+
+        return self.generate(X, **kwargs)
+
+    def generate(self, X: t.Any, **kwargs):
```

Comment:
Shouldn't this be `predict` -- like in `OpenAI`?

Reply:
Yes, you reminded me, I'll take care of it

Reply:
Done